### ISSUE TITLE
documentation clarification regarding tidy eval differences between mutate and select

### ISSUE THREAD

**Issue by mkoohafkan** [tidyverse/dplyr#4000]
I struggled with some tidy eval semantics today and am looking for some clarification on the reasoning behind differences in how tidy evaluation works for e.g. `select` vs `mutate`. I spent a lot of time looking at the [programming with dplyr vignette](https://cran.r-project.org/web/packages/dplyr/vignettes/programming.html), and the [tidy eval with rlang cheatsheet](https://github.com/rstudio/cheatsheets/blob/master/tidyeval.pdf), but I'm still struggling to understand *why* these differences exist.

The context of the problem is using `dplyr` inside a package, where some manipulation of data frames with fixed, known column names may occur (perhaps data is downloaded from a web service and then additional manipulation is done before serving it to the user). Here's a simplified example of my confusion:

```
library(dplyr)
mtcars %>% select(mpg)                # interactive approach, causes warnings in package
mtcars %>% select(.data$mpg)      # works without warnings if rlang::.data is imported
mtcars %>% select(!!"mpg")           # just works!

mtcars %>% mutate(mpg2 = mpg * 2)            # interactive approach, causes warnings in package
mtcars %>% mutate(mpg2 = .data$mpg * 2)  # works without warnings if rlang::.data is imported
mtcars %>% mutate(mpg2 = !!"mpg" * 2)      # does not work, !!"mpg" is interpreted as character rather than a column
```

**Why does the `!!"varname"` approach work with `select` but not `mutate`?** After messing around a lot and eventually finding the cheatsheet, I found that this approach with `rlang::parse_expr` also works:

```
mtcars %>% select(!!parse_expr("mpg"))
mtcars %>% mutate(mpg2 = !!parse_expr("mpg") * 2)
```

However, `parse_expr` is not mentioned in the vignette.

I have moved to using the `.data` approach since it is the safest approach and avoids ambiguity, but I am still curious about the technical reasons behind why `!!"varname"` syntax works with `select` but not `mutate`. 

Or, maybe this is all just a bug in `select`, since this works (but it shouldn't?):

```
mtcars %>% select("mpg")
```

**Comment by dpprdan** [comment]
> Why does the `!!"varname"` approach work with `select` but not `mutate`?

See the last example in the `?select()` help file:

> For convenience it also supports strings and character vectors. This is unlike other verbs where strings would be ambiguous.

I suppose, your opening this issue shows that this exception could be made more explicit, though.

**Comment by mkoohafkan** [comment]
Ok, thanks---so really, `!!"varname"` never worked like I thought it did---it always evaluates to a string (rather than the column itself), but it just so happens that `select` works with strings.

I think an example of `parse_expr` shortly before the `.data` examples would be more helpful than emphasizing the unique behavior of `select`. My misunderstanding came about because I was trying to replace `select_` and `mutate_` calls where I used strings) with tidy eval.

**Comment by yutannihilation** [comment]
For the differences between select semantics and mutate semantics, this intro is very helpful :)

https://dplyr.tidyverse.org/articles/dplyr.html#selecting-operations

**Comment by mkoohafkan** [comment]
Thanks @yutannihilation, but this documentation actually makes me more confused. Consider the following statement in the linked document:

>Mutate semantics are quite different from selection semantics. Whereas select() expects column names or positions, mutate() expects column vectors.

Does that mean the statement

```
mtcars %>% select(.data$mpg)
```

Is **not** the correct syntax, since `select` semantics should evaluate to a column *position* rather than a column *vector*?

**Comment by yutannihilation** [comment]
Ah, agreed `$` seems subsetting operation, but it's actually not; that `.data$mpg` is evaluated to a position. Sorry I can't explain better...

**Comment by mkoohafkan** [comment]
Ah, ok. So in the example `select` statement the `.data` variable looks something like

```
list(mpg = 1L, cyl = 2L, disp = 3L, ...) # etc.
```

Whereas in the `mutate` statement `.data` is the complete dataframe.

**Comment by hadley** [comment]
Yeah, exactly.

**Comment by hadley** [comment]
I think this is mostly fixed by #4684. I'm also going to revamp all the tidy eval vignette, and selection vs action semantics is now part of our standard text, so there will also be more details there.
